sklearn/feature_selection/_univariate_selection.py

```diff
@@ -288,6 +290,189 @@ def chi2(X, y):
     return _chisquare(observed, expected)
 
 
+def _get_entropy(prob):
+    t = np.log2(prob)
+    t[~np.isfinite(t)] = 0
+    return np.multiply(-prob, t)
+
+
+def _info_gain(X, y, aggregate, ratio=False):
+
+    def _a_log_a_div_b(a, b):
+        with np.errstate(invalid="ignore", divide="ignore"):
+            t = np.log2(a / b)
+        t[~np.isfinite(t)] = 0
+        return np.multiply(a, t)
+
+    f_count, c_count, fc_count, total = _get_fc_counts(X, y)
+
+    f_prob = f_count / f_count.sum()
+    c_prob = c_count / c_count.sum()
+    fc_prob = fc_count / total
+
+    c_f = _a_log_a_div_b(fc_prob, c_prob * f_prob)
+    nc_nf = _a_log_a_div_b(1 - f_prob - c_prob + fc_prob, (1 - c_prob) * (1 - f_prob))
+    c_nf = _a_log_a_div_b((c_count - fc_count) / total, c_prob * (1 - f_prob))
+    nc_f = _a_log_a_div_b((f_count - fc_count) / total, (1 - c_prob) * f_prob)
+
+    scores = c_f + nc_nf + c_nf + nc_f
```

Comment:
I think that I would prefer `_info_gain` to return this score and

have the ratio below done in the `info_gain_ratio` and finally have a function to that could be called twice to just make the reduction.


```python
def _info_gain(X, y):
    # probably the name of the function should be better.
    ...
    return scores, c_prob

def info_gain(X, y, aggregate=np.max):
    return aggregate.reduce(_info_gain(X, y)[0], axis=0)

def info_gain_ratio(X, y, aggregate=np.max):
    scores, c_prob = _info_gain(X, y)
    with np.errstate(invalid="ignore", divide="ignore"):
        scores /= (entropy(c_prob) + entropy(1 - c_prob))
    return aggregate.reduce(scores, axis=0)
```